hw07_file_copying/copy.go

```diff
@@ -2,14 +2,102 @@ package main
 
 import (
 	"errors"
+	"io"
+	"log"
+	"os"
+
+	"github.com/cheggaaa/pb/v3"
 )
 
+const bufSize = 1024
+
 var (
 	ErrUnsupportedFile       = errors.New("unsupported file")
 	ErrOffsetExceedsFileSize = errors.New("offset exceeds file size")
 )
 
+var readListeners []func(n int)
+
+func readEvent(n int) {
+	for _, listener := range readListeners {
+		listener(n)
+	}
+}
+
+func initBar(total int64) *pb.ProgressBar {
+	bar := pb.Start64(total)
+	readListeners = append(readListeners, func(n int) {
+		bar.Add(n)
+	})
+	return bar
+}
+
 func Copy(fromPath, toPath string, offset, limit int64) error {
-	// Place your code here.
-	return nil
+	wFile, err := os.OpenFile(toPath, os.O_CREATE|os.O_RDWR|os.O_TRUNC, 0o644)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := wFile.Close(); err != nil {
+			log.Fatal(err)
+			return
+		}
+	}()
+	rFile, err := os.Open(fromPath)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := rFile.Close(); err != nil {
+			log.Fatal(err)
+			return
+		}
+	}()
+	stat, err := rFile.Stat()
+	if err != nil {
+		return err
+	}
+	size := stat.Size()
+	if size == 0 {
+		_, err := rFile.Read(make([]byte, 2))
+		if !errors.Is(err, io.EOF) {
+			return ErrUnsupportedFile
+		}
+	}
+	if size < offset {
+		return ErrOffsetExceedsFileSize
+	}
+	bar := initBar(size)
+	defer bar.Finish()
+	return copyResource(rFile, wFile, offset, limit)
+}
+
+func copyResource(r io.ReadSeeker, w io.Writer, offset, limit int64) error {
+	var reed int64
```

Comment:
```suggestion
	var read int64
```
не понимаю что делает эта переменная. Можете её по другому как то назвать?